Splitting the SEPA tokens for other methods setting

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes STRIPE-693
P2 peNR48-1Jq-p2

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

Following the discussion in peNR48-1Jq-p2, in this PR, I am splitting the "Enable SEPA Tokens for other methods" setting into separate settings for both iDEAL and Bancontact. I am also updating some of the copy based on the suggestion.

This setting was first introduced in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3819. It allows shoppers to purchase using iDEAL or Bancontact, and saving the payment method as a SEPA Direct Debit token. You can read more about this Stripe behavior in the following links:
- https://docs.stripe.com/billing/subscriptions/bancontact
- https://docs.stripe.com/payments/ideal/set-up-payment

Settings preview:
<img width="770" height="160" alt="Screenshot 2025-08-28 at 15 13 58" src="https://github.com/user-attachments/assets/434a2171-0570-4e16-b942-9a9e3a05b260" />

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`update/split-sepa-tokens-for-other-methods-setting`)
- Connect your Stripe account
- Set your store currency to EUR
- Enable Bancontact and iDEAL
- Confirm you can see both settings and they are enabled by default
- Confirm you can disable/enable the settings and that by reloading the page, the change is persisted
- As a shopper, with the first setting enabled, try to add a new iDEAL method
- Confirm the method is saved
- Disable the setting and confirm you are not able to save the method anymore
- Repeat the test steps for Bancontact

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Splitting SEPA tokens for other methods setting

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Readme and changelog entries

wjrosa committed:
Fix tests

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

diegocurbelo committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

diegocurbelo committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

annemirasol reviewed (commented):
Question about the naming for the iDEAL setting. I'm seeing that we're using `is_sepa_tokens_for_other_methods_enabled`, while we're using a specific name for Bancontact (`is_sepa_tokens_for_bancontact_enabled`). Why are we not renaming it to be iDEAL-specific?

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa commented:
> Question about the naming for the iDEAL setting. I'm seeing that we're using `is_sepa_tokens_for_other_methods_enabled`, while we're using a specific name for Bancontact (`is_sepa_tokens_for_bancontact_enabled`). Why are we not renaming it to be iDEAL-specific?

Good question, Anne. The only reason is to avoid the need for a new migration script (to convert the existing option into the new ones). That was the simplest solution. But I am happy to go the other way if you believe it is better.

diegocurbelo reviewed (commented):
The changes look good, but I agree that renaming `...for_other_methods` to `...for_ideal` would make the code easier to follow/understand long term, even if we need an initial migration.

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Renaming the ideal setting properly + adding a migration script

wjrosa commented:
Alright, folks! I renamed all references to the old setting name, updated the implementation, deprecated public methods on the backend, and added a migration script in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4633/commits/2e9b1d96651dd3866866ab9ac8c714ffb821a516

wjrosa committed:
Fix include

annemirasol commented on includes/admin/stripe-settings.php:
```diff
@@ -226,10 +226,18 @@
 			'desc_tip'    => true,
 		],
 		'sepa_tokens_for_other_methods'       => [
```
Should we rename this as well? It is still `sepa_tokens_for_other_methods`

annemirasol commented on client/settings/payments-and-transactions-section/index.js:
```diff
@@ -99,11 +104,23 @@ const PaymentsAndTransactionsSection = () => {
 					checked={ isSepaTokensForOtherMethodsEnabled }
 					onChange={ setIsSepaTokensForOtherMethodsEnabled }
```
Should we rename these as well?

annemirasol commented on client/settings/payments-and-transactions-section/index.js:
```diff
@@ -19,6 +19,7 @@ import {
 	useEnabledPaymentMethodIds,
 	useIsShortAccountStatementEnabled,
 	useSepaTokensForOtherMethods,
```
I think we need to rename this too -- we renamed the settings hook in `client/data/settings/hooks.js`.

annemirasol reviewed (commented):
Thanks for the renames and adding a migration script. Found a few more renames we might have missed.

wjrosa committed:
Replacing missing occurrences

wjrosa commented:
Oops! Thanks for the heads up! Replaced all the remaining occurrences in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4633/commits/11b54b8f4d539fada64eade07fadfac75ff8b7d0

diegocurbelo commented on includes/migrations/class-sepa-tokens-for-other-methods-settings-update.php:
```diff
@@ -0,0 +1,50 @@
+			return;
+		}
+
+		$value = $stripe_gateway->get_option( 'sepa_tokens_for_other_methods', 'no' );
```
[nit] We can't remove the `sepa_tokens_for_other_methods`, as we want to keep the current behavior in case of a rollback/downgrade... but maybe we should create a follow-up issue (or add a TODO here) to remove it in 10.2 (L-2) so we don't keep unused options in the settings

wjrosa replied:
Good idea. Added it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/4633/commits/ff32196943340d1a4abbd041c0f324a1e82bc105

diegocurbelo reviewed (approved):
Thanks for the changes... it looks good and tests well 🚢

annemirasol reviewed (approved):
LGTM!

wjrosa committed:
Merge branch 'develop' into update/split-sepa-tokens-for-other-methods-setting

wjrosa committed:
Adding TODO comment about deprecating the old setting

wjrosa committed:
Changelog and readme entries
